### DISCUSSION TITLE
Optimize image size by DPI

### DISCUSSION THREAD

**Discussion by wbijker** [QuestPDF/QuestPDF#396] [Q&A]
I like to reduce PDF size by resizing images. The problem is the images are dynamic in resolution and I don't necessarily know the available container space to do the resizing myself. Maybe a valid scenario for #36.  .

Currently I use the canvas as a solution:
```c#
public static void ImageScaled(this IContainer container, SKImage source, int dpi, ImageCrop imageCrop) 
{
    container.Canvas((canvas, space) =>
            {
                // when RasterDpi is set to 72, one PDF point corresponds to one pixel.
                // space measure in PDF points.
                var spacePx = new Size(space.Width * (dpi / 72f), space.Height * (dpi / 72f));
                // omitted for brevity
                canvas.DrawImage(...);
            });
}
```

But it looks like canvas elements can't be shared for multiple pages. ie. If.
```c#
page.Size(PageSizes.A4.Landscape());

                page.Background()
                    .ExtendVertical()
                    .Element(c => { c.ImageScaled("background.png"); });
```

If I then have 10 pages of content the output size tells me the background.png is repeated 10x times.

My train of thought:
1.  If available space is known, I can do the DPI calculation myself and use the normal IContainer.Image image. Or I should always use explicit dimensions when working with raster data.
2. Maybe there exists a way to work with raster images DPI in Skia? Like to specify a DPI and paint as normal.
3. Maybe there's a way to let the canvas element be shared across pages?

**Comment by MarcinZiabek** [comment]
As far as I know, SkiaSharp automatically optimizes images based on ducument's DPI. However, you may want to consider changing image quality from loseless (png) to more compressed form (jpg). You can do it with [the DocumentMetadata.ImageQuality option](https://www.questpdf.com/concepts/document-metadata#document-metadata).

**Comment by wbijker** [comment]
```c#
var document = Document.Create(doc =>
            { 
                doc.Page(page =>
                {
                    page.Size(PageSizes.A4.Landscape());
                    page.Content().Extend().Image("pdf/citrus.png", ImageScaling.Resize);
                });
            });
            document.WithMetadata(new DocumentMetadata()
            {
                ImageQuality = 75,
                RasterDpi = 72
            });
```

RasterDPI is respected when generating an image, but not PDF.

**Comment by DawidRobertNowak** [comment]
Hello @wbijker,

We are closing this thread due to inactivity. We apologize for the delayed response.

We're happy to share that both concerns you raised have since been addressed:

- **DPI-based image optimization**: QuestPDF now automatically resizes images to match a target DPI based on the available container space — no need to calculate dimensions manually. You can set this per image with `.WithRasterDpi()` or globally via `DocumentSettings.ImageRasterDpi`.

- **Shared images across pages**: Images loaded from file paths are now automatically cached and embedded only once in the PDF, so a background image used across 10 pages will no longer inflate the file size.

Additionally, you can control compression quality per image with `.WithCompressionQuality()` or globally via `DocumentSettings.ImageCompressionQuality`.

Full documentation is available here:
https://www.questpdf.com/api-reference/image/optimization.html
https://www.questpdf.com/api-reference/image/shared.html

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
